### DISCUSSION TITLE
Scheduling tasks with different frequencies and finding first feasible solution

### DISCUSSION THREAD

**Discussion by JackBogart** [google/or-tools#3887] [CP-SAT questions]
Hey there,
I have a bit of a complex problem that I've been having trouble modelling. I want to preface that I'm very new to constraint programming, I'm still in college and not experienced.

My problem is this: Given a list of tasks that have a duration and a frequency I want to find a feasible schedule that fits all the tasks. Frequency can be any ratio that is greater than 0, and the schedule is confined to a fixed length that the tasks fit in. 

For example if I have as my list of tasks: {(30s, 1/2), (1, 2/1), (15, 1/2)} with a schedule length of 60 an example of a solution would be:
Loop 1: 30s 1s 1s
Loop 2: 1s 1s 15s

This example is just one of the many possible schedules that could be made. Any help would be appreciated, even just a general direction on how to attack this. I've looked at the scheduling problems, assignment problems, and even tried approaching it like a 0-1 multiple knapsack but the frequency is the main problem I'm having. Is CP-SAT a good tool for this problem?

**Comment by lperron** [comment]
The biggest challenge is to find how many period you need to consider. With various ratios, it could explore. 3/7 and 1/4...

After that, you sum potential and mandatory durations in each period and it could be easy.

**Reply by JackBogart** [reply]
Do you have any recommendations on an example to look at that's similar?

Also yeah the different ratios could get unruly, the current way it's handled is restricting the denominator of the frequency to 1, 2, 3, 4, 6, and 8. This makes it so taking the LCM of the all the task frequencies doesn't result in a large amount of periods and is implicitly capped at 24.
